### PR TITLE
Support generating a syntax tree for the redacted content 

### PR DESCRIPTION
**Pull request by bethanyaconnor**
Adds a plugin to find restoration locations and a plugin to merge the syntax trees for the source and redacted contents.

This review will be easier to view each of the plugins on their own. `restore.js` is totally rewritten, with a lot of that functionality moving to `findRestorations` so the diffs aren't very useful.

Related: https://github.com/code-dot-org/redactable-markdown/pull/76, https://github.com/code-dot-org/remark-plugins/pull/17

### CODE REVIEW

File: src/restore.js [hunk 1]
```diff
@@ -25,156 +26,64 @@ const visit = require("unist-util-visit");
...
-    if (!this.Parser) {
-      return;
+  function blockVisitor(node, index, parent) {
+    const redactedData = redactions[node.redactionIndex];
+    if (!redactedData || redactedData.type !== "blockRedaction") {
+      const nodes = [
+        {type: "rawtext", value: "[][" + node.redactionIndex + "]"},
+        ...node.children,
+        {type: "rawtext", value: "[/][" + node.redactionIndex + "]"}
+      ];
```
Comment (Hamms) [inline src/restore.js]: why do this instead of leaving it as a `blockRestoration` node?

File: src/restore.js [hunk 2]
```diff
@@ -25,156 +26,64 @@ const visit = require("unist-util-visit");
...
-      // if in normal (ie non-silent) mode, consume the token and produce a
-      // render
-      const subvalue = value.slice(startIndex, endIndex);
-      const children = this.tokenizeBlock(subvalue, eat.now());
-      const add = eat(blockOpen + subvalue + blockClose);
-      return restorationMethod(add, redactedData, content, children);
+    const restorationMethod = restorationMethods[redactedData.redactionType];
+    if (restorationMethod) {
+      const restored = restorationMethod(redactedData, node.content);
+      Object.assign(node, restored);
```
Comment (Hamms) [inline src/restore.js]: why this instead of `parent.children.splice(index, 1, restored)`?

**Comment by bethanyaconnor** [comment]
I separated out a compiler plugin to render missed restorations. I didn't love the idea of it being in the same file as `parseRestorations` because the plugins are used in different places, so I kept them separate for now.

I also renamed/refactored `parseRestorations`. LMK what you think!

File: src/restore.js (hunk 1 above)
Comment (bethanyaconnor) [reply src/restore.js]: Changed!

File: src/restore.js (hunk 2 above)
Comment (bethanyaconnor) [reply src/restore.js]: Changed. Splicing in the node is definitely better.

File: test/restoration.test.js [hunk 3]
```diff
@@ -3,12 +3,20 @@ const path = require("path");
...
 
 fs.readdirSync(path.resolve(__dirname, "data")).forEach(testCategory => {
   const dataPath = path.resolve(__dirname, "data", testCategory);
 
   const plugin = require(path.resolve(dataPath, "plugin.js"));
+  let restorationMethods;
+  if (plugin.length) {
+    restorationMethods = plugin
+      .map(p => p.restorationMethods)
+      .reduce((acc, val) => Object.assign({}, acc, val), {});
```
Comment (Hamms) [inline test/restoration.test.js]: should this logic be exposed somehow?

**Comment by Hamms** [comment]
My thinking is that although the parser logic is only ever used when `this.Parser` is present and the compiler logic is only ever used when `this.Compiler` is present, they're both still dealing with the same content concept - namely, restoration nodes - and should be kept in the same place. In particular, I like the idea of there being a plugin that contains everything a process needs to know to deal with restoration nodes, and you can just stick that plugin in a `unified` processor without needing to care about exactly which part of the process that particular processor is handling.

In that same vein, what do you think about renaming things to make it a little more explicit what the `restore` operation is actually doing? Right now, we have:

- `redact` - a plugin to deal with redactions
- `restore` - a plugin to merge a tree with redactions with another one with restorations
- `parseRestorations` - a plugin to parse restorations
- `renderRestorations` - a plugin to render restorations

What if instead we did something like:

- `redact` (or `redaction`?) - a plugin to handle redactions
- `restore` (or `restoration`?)- a plugin to handle restorations
- `mergeRedactionAndRestoration` (or `redaction-restoration-to-regular`, to mirror https://github.com/syntax-tree/mdast-util-to-hast? or `redact/restore-remark`, to mirror https://github.com/remarkjs/remark-rehype?) - a plugin to merge a tree with redactions and another with restorations to a regular ol' remark tree

**Comment by Hamms** [comment]
As an aside, would you also update https://github.com/code-dot-org/remark-redactable/blob/master/docs/mdast-nodes.md to document the `Restoration` nodes?

**Pull request merged**